### ISSUE TITLE
How to predict with trained model

### ISSUE THREAD

**Issue by lich0821** [yuanenming/mb-str#9]
Hello Enming, I could reproduce the experiment results of your paper. However, when I used the trained model to make inference the result seemed not as good as experiments.

I think I made some mistakes but I couldn't find it. Could you help take a look? Thank you.

Here's how I predict with trained model:
```py
N = 20
backbone = BERT(max_len=50, num_items=22734, n_layer=2, n_head=2, n_b=4,
                d_model=16, dropout=0.1, battn=True, bpff=True, brpb=True,)
model = RecModel(backbone, b_head=True)
model_path = "logs/yelp/full/lightning_logs/version_0/checkpoints/epoch=29-step=299.ckpt"
checkpoint = torch.load(model_path, map_location=torch.device("cpu"), weights_only=False)
model.load_state_dict(checkpoint["state_dict"])
model.eval()

seq, b_seq = [padded data], [padded data]
input_ids = torch.tensor([seq])
b_seq = torch.tensor([b_seq])

with torch.no_grad():
    outputs = model(input_ids, b_seq)
    last_outputs = outputs[:, -1, :]
    last_b_seq = b_seq[:, -1]
    logits = model.head(last_outputs, last_b_seq)
    rank = (-logits).argsort(dim=1)
    rank = rank.cpu().numpy()[0]
    print(f"rank@{N}: {rank[:N]}")
```

**Comment by yuanenming** [comment]
Hi, I have checked the code you provided. There are several concerns:
1. Check if the model is fully-trained because I see the checkpoint is recorded at epoch 29. the yelp dataset is pretty small. It is likely the model is still under-trained.
2. The model inference should be performed in a BERT style, which means you should fill in with a `[MASK]` token in the item sequence but leave the behavior sequence unmasked. Please check if it is performed under this framework.

**Comment by lich0821** [comment]
Thank you for your reply.
1. I think it's fully trained as I set `max_epochs` to 300 and it finished without errors. But even it's not full-trained, it doesn't make any difference because I use the same trained model to perfrom valid and test. Is it right?
2. I didn't make it clear. In the padded seq data, I put an extra `0` at the end; and put a target behavior to the end of padded b_seq.

**Comment by yuanenming** [comment]
OK, I got it.

How do you evaluate the prediction when inference? It seems like you print the rank of the ground-truth items. 

Note that the evaluation metrics used in the experiment are calculated by the ground-truth item vs. 99 randomly sampled non-interactive items. You may get very poor performance if you treat all items as negative samples.

**Comment by lich0821** [comment]
I got top20 predicted items, and then compared with items that the user behaved.

For example, data for user U:
```py
seq = [76,22,35,126,90,37,11,130,27,151,154,107,102,57,75,58,83,70,91,143]
b_seq = ["pos","pos","pos","pos","pos","pos","pos","neg","pos","pos","neg","pos","pos","tip","pos","pos","pos","neutral","neg","neg"]
```

I'll make input as:
```py
# Pad and add [MASK] to the end
seq = [0, 0, ..., 76, 22, 35, 126, 90, 37, 11, 130, 27, 151, 154, 107, 102, 57, 75, 58, 83, 70, 91, 143, 0]  
# {'neutral': 1, 'neg': 2, 'pos': 3, 'tip': 4}
# Pad and add target to the end
b_seq = [0, 0, ..., 3, 3, 3, 3, 3, 3, 3, 2, 3, 3, 2, 3, 3, 4, 3, 3, 3, 1, 2, 2, 3]  # Predict 3 (pos)
```

Then get the top20 predicted items and check if these items are in those that user U rated `pos`.

**Comment by lich0821** [comment]
Well, do MBSTR filter items that have showed in training set? This might be the reason.😵‍💫

**Comment by yuanenming** [comment]
> Note that the evaluation metrics used in the experiment are calculated by the ground-truth item vs. 99 randomly sampled non-interactive items. You may get very poor performance if you treat all items as negative samples.

Again, I think the problem is about the evaluation metrics. If you treat all item sets as negative samples (more than 20k items), and expect the ground truth item ranked in the top 20. It is somehow equal to top-1 vs 1k negative. It is a much hard task than the one used in our experiments.

**Comment by JuweiChan** [comment]
> Thank you for your reply.
> 
> 1. I think it's fully trained as I set `max_epochs` to 300 and it finished without errors. But even it's not full-trained, it doesn't make any difference because I use the same trained model to perfrom valid and test. Is it right?
> 2. I didn't make it clear. In the padded seq data, I put an extra `0` at the end; and put a target behavior to the end of padded b_seq.

I think the [MASK] token should be "num_items+1" rather than "0" according to the code?

**Comment by lich0821** [comment]
> > Thank you for your reply.
> > 
> > 1. I think it's fully trained as I set `max_epochs` to 300 and it finished without errors. But even it's not full-trained, it doesn't make any difference because I use the same trained model to perfrom valid and test. Is it right?
> > 2. I didn't make it clear. In the padded seq data, I put an extra `0` at the end; and put a target behavior to the end of padded b_seq.
> 
> I think the [MASK] token should be "num_items+1" rather than "0" according to the code?

Well, that's important. @yuanenming Enming, could you confirm this? Thank you~

**Comment by yuanenming** [comment]
https://github.com/yuanenming/mb-str/blob/6ead831ea4e7073d69d576008be2cb917c9c8ccf/src/dataloaders/rec_dataloader.py#L83

Yes, I have checked the code.

**Comment by lich0821** [comment]
That's the problem. Thank you @JuweiChan and @yuanenming
